packages/foundation/src/createComponent.tsx

```diff
@@ -63,15 +75,69 @@ export function createComponent<
     }
 
     const theme = componentProps.theme || settings.theme;
+    const tokens = _resolveTokens(componentProps, theme, options.tokens, settings.tokens, componentProps.tokens) as any;
+    let styles;
+
+    const finalStyles: { [key: string]: string | undefined } = {};
+
+    if (!options.disableCaching) {
+      // We get the entry in the memoized classNamesMap for the current component or create one if it doesn't exist.
+      const displayName = options.displayName || view.name;
+      if (!memoizedClassNamesMap.hasOwnProperty(displayName)) {
+        memoizedClassNamesMap[displayName] = { map: {} };
+      }
+
+      // Memoize based on the tokens definition.
+      let current = memoizedClassNamesMap[displayName];
+      const tokenKeys = Object.keys(tokens).sort();
+      for (const key of tokenKeys) {
+        let nextToken = tokens[key];
+        if (nextToken === undefined) {
+          nextToken = '__undefined__';
+        }
+        if (!current.map.hasOwnProperty(nextToken)) {
+          current.map[nextToken] = { map: {} };
+        }
+        current = current.map[nextToken];
+      }
 
-    const tokens = _resolveTokens(componentProps, theme, options.tokens, settings.tokens, componentProps.tokens);
-    const styles = _resolveStyles(componentProps, theme, tokens, options.styles, settings.styles, componentProps.styles);
+      // Memoize based on the base styling of the component (i.e. without user specified props).
+      const defaultStyles = _resolveStyles(componentProps, theme, tokens, options.styles, settings.styles) as any;
```

Comment:
How much time is this taking up? I'm wondering if there is any way to memoize this as well or if the cost/benefit would just break even or flip. (I seem to remember the merging of objects on every pass consuming not in significant amounts of time.)

Reply:
I've done some changes so that this is now just called the first time per component as needed.